src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderValidationTests.cs

```diff
@@ -146,6 +146,131 @@ public void GetService_DoesNotThrow_WhenGetServiceForPolymorphicServiceIsCalledO
+            serviceCollection.AddKeyedTransient<ServiceKeyWithStronglyTypedArgument>(KeyedService.AnyKey);
+            using var serviceProvider = serviceCollection.BuildServiceProvider(new ServiceProviderOptions
+            {
+                ValidateScopes = true,
+                ValidateOnBuild = validateOnBuild
+            });
+
+            // Act
+            var actual = serviceProvider.GetKeyedService<ServiceKeyWithStronglyTypedArgument>(42);
+
+            // Assert
+            Assert.Equal(42, actual.Key);
+            Assert.Throws<InvalidOperationException>(() => serviceProvider.GetKeyedService<ServiceKeyWithStronglyTypedArgument>("Hello"));
+        }
+
+        private class ServiceKeyWithStronglyTypedArgument
+        {
+            public int Key { get; set; }
+
+            public ServiceKeyWithStronglyTypedArgument([ServiceKey] int key)
+            {
+                Key = key;
+            }
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void BuildServiceProvider_AnyKey_ServiceKeyWithObjectTypedArgument(bool validateOnBuild)
+        {
+            // Arrange
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddKeyedTransient<ServiceKeyWithObjectTypedArgument>(KeyedService.AnyKey);
+            using var serviceProvider = serviceCollection.BuildServiceProvider(new ServiceProviderOptions
+            {
+                ValidateScopes = true,
+                ValidateOnBuild = validateOnBuild
+            });
+
+            // Act
+            var actualInt = serviceProvider.GetKeyedService<ServiceKeyWithObjectTypedArgument>(42);
+            var actualString = serviceProvider.GetKeyedService<ServiceKeyWithObjectTypedArgument>("hello");
+
+            // Assert
+            Assert.Equal(42, actualInt.Key);
+            Assert.Equal("hello", actualString.Key);
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void BuildServiceProvider_ServiceKeyWithObjectTypedArgument(bool validateOnBuild)
+        {
+            // Arrange
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddKeyedTransient<ServiceKeyWithObjectTypedArgument>(42);
+            serviceCollection.AddKeyedTransient<ServiceKeyWithObjectTypedArgument>("hello");
+            using var serviceProvider = serviceCollection.BuildServiceProvider(new ServiceProviderOptions
+            {
+                ValidateScopes = true,
+                ValidateOnBuild = validateOnBuild
+            });
+
+            // Act
+            var actualInt = serviceProvider.GetKeyedService<ServiceKeyWithObjectTypedArgument>(42);
+            var actualString = serviceProvider.GetKeyedService<ServiceKeyWithObjectTypedArgument>("hello");
+            var notFound = serviceProvider.GetKeyedService<ServiceKeyWithObjectTypedArgument>(false);
```

Comment:
>GetKeyedService [](http://example.com/codeflow?start=43&length=15)

what would be the behavior when passing `null` as a parameter? will `GetKeyedService` return null? or will return `ServiceKeyWithObjectTypedArgument` object with `Key = null`?

Reply:
Passing `null` means that this is not a "keyed service" so it just uses the type which is `ServiceKeyWithObjectTypedArgument` and since that was not registered, it returns `null`.

These checks are really just verifying that `[ServiceKey] object key` is called with the correct key and no unexpected exceptions occur.